Upgrade to linkerd 2.10.2 (#41)
giantswarm/linkerd2-app: README.md

    tls:
      crtPEM: |
        <contents of the issuer.crt file>
      keyPEM: |
        <contents of the issuer.key file>
identityTrustAnchorsPEM: |
  <contents of the ca.crt file>
```

- Download the [default values.yaml file](helm/linkerd2-app/values.yaml) and save it as `my-linkerd-values.yaml` to create a [user configuration](https://docs.giantswarm.io/app-platform/app-configuration/). Review the values in the file. With the default values, linkerd will be installed in High-Availability mode and with CNI plugin enabled.

### Step 2: Deploy Linkerd

We recommend deploying the app by applying an `App` CR (Custom Resource) onto your management cluster. Use the [`kubectl gs`](https://docs.giantswarm.io/ui-api/kubectl-gs/) plugin to generate a valid `App` CR with command:

```bash
kubectl gs template app \
  --catalog giantswarm \
  --name linkerd2-app \
  --namespace linkerd \
  --cluster <your-cluster-id>  \
  --version 0.6.0 \
  --user-configmap my-linkerd-values.yaml \
  --user-secret my-linkerd-certificates.yaml \
  --namespace-labels "linkerd.io/is-control-plane=true,config.linkerd.io/admission-webhooks=disabled,linkerd.io/control-plane-ns=linkerd" \

>   --namespace-annotations "linkerd.io/inject=disabled > linkerd-manifest.yaml"

Author: ```suggestion
  --namespace-annotations "linkerd.io/inject=disabled" > linkerd-manifest.yaml
```

```

The final `App` CR should look like this:

```yaml
...
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: linkerd2-app
  namespace: <your-cluster-id>
spec:
  catalog: giantswarm
  kubeConfig:
    inCluster: false
  name: linkerd2-app
  namespace: linkerd
  namespaceConfig:
    annotations:
      linkerd.io/inject: disabled
    labels:
      config.linkerd.io/admission-webhooks: disabled
      linkerd.io/control-plane-ns: linkerd
      linkerd.io/is-control-plane: "true"
  userConfig:
    configMap:
      name: linkerd2-app-userconfig-<your-cluster-id>
      namespace: <your-cluster-id>
    secret:
      name: linkerd2-app-userconfig-<your-cluster-id>
      namespace: <your-cluster-id>
  version: 0.6.0
```

Please note, that for security reasons Giant Swarm by default forbids the usage of

> `emptyDir` volumes as storage for pods. Linkerd needs this functionality to deploy
> linkerd itself, but also to run any deployment that is going to be included in

Author: ```suggestion
`emptyDir` volumes as storage for pods. Linkerd needs this functionality to deploy the
linkerd control plane itself, but also to run the proxy containers in any deployment that is going to be included in
```

the service mesh. Enabling `emptyDir` volumes poses a risk that a pod will create
such big `emptyDir` that the underlying cluster node will run out of disk space.
If you're OK with this potential issue, the easiest way to allow for `emptyDir`
for all the deployments is to edit the default PSP of your cluster by running the
command:

```bash
kubectl edit psp restricted
```

then finding the `spec.volumes` list and appending `emptyDir` value to the list. Then, write the file and exit your editor.

> To finally deploy linkerd, run

Author: The chart includes an own PSP, so for the bare app to be installed, the `restricted` psp does not need to be edited, so IMO this `finally` could be a little confusing

# linkerd2-app chart

This chart is based on the official linkerd2 helm chart.

## Requirements

- you can install only one release of this chart per kubernetes cluster.
- it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
  - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.
- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes

>   - the injected linkerd-proxy container requires an `EmptyDir` volume

Author: sorry to be the party-pooper again, its two volumes.. one for the `proxy-init` initContainer and one for the `proxy` sidecar container

check https://github.com/giantswarm/linkerd2-app/blob/subtree-2.10/helm/partials/templates/_volumes.tpl

```suggestion
  - the injected linkerd-proxy container requires an `EmptyDir` volumes
```

>   - the injected linkerd-proxy container requires an `EmptyDir` volumes

Author: Ok I am a little doofus :crab:  @ced0ps please help :P

```suggestion
  - the injected linkerd-proxy container and proxy initContainer require an `EmptyDir` volume
```

      name: linkerd2-app-userconfig-<your-cluster-id>
      namespace: <your-cluster-id>
    secret:
      name: linkerd2-app-userconfig-<your-cluster-id>
      namespace: <your-cluster-id>
  version: 0.6.0

```

When installing via the Giant Swarm web UI, you'll need to apply the above labels and annotations manually.

### Step 4: After deployment

- Disable proxy injections for pods running in the `kube-system` namespace by applying certain labels and annotations

```bash
kubectl annotate namespace kube-system linkerd.io/inject=disabled
kubectl label namespace kube-system config.linkerd.io/admission-webhooks=disabled
```

- Optional but recommended: You can use the `linkerd` cli as usual with this app as we're using the default namespaces. (`linkerd` and `linkerd-cni`). You can download it from the [linkerd release page](https://github.com/linkerd/linkerd2/releases/tag/stable-2.10.2).

- We strongly recommend installing the `linkerd viz` extension using the [`linkerd`](#usage-with-linkerd-cli) by executing

```bash
linkerd viz install | kubectl apply -f -
```

After installation, you can open the dashboard by executing

```
linkerd viz dashboard
```

## Mesh your workloads

> After installation, linkerd looks for a `linkerd.io/inject: enabled` annotation on `Namespaces` or other workload resources. Adding this annotation to your workload namespaces will trigger automatic proxy container injection to your pods. More information on proxy injection can be found on the ["Automatic Proxy Injection" page](https://linkerd.io/2.10/features/proxy-injection/) in the upstream documentation.

Reviewer: Maybe let's add info about how to use `namespaceConfiguration` in App CRs to get the same effect for apps deployed with app-platform?

Author: added :heavy_check_mark:

**Attention**: Proxy containers are using `EmptyDir` volumes for storing ephemeral data, so all of your workload pods meshed by linkerd require a `PodSecurityPolicy` which allows use of `EmptyDir` volumes.

## Installing without the CNI plugin

In order to install this app without the CNI plugin, you'll need to specify `cniEnabled: false` in your user configuration.

Be aware that running without the CNI plugin, proxy containers will run as `root` and will require `NET_ADMIN` and `NET_RAW` capabilities.

## Troubleshooting

### App installation fails when I specify a namespace name other than `linkerd`

The `linkerd` namespace name is assumed in several parts of the chart, so we restricted installation to namespaces named `linkerd` only.

### Proxy containers fail to start

> Your workload must be able to create `EmptyDir` volumes, so you'll need to create a `PodSecurityPolicy` allowing the creation of `EmptyDir` volumes.

Reviewer: We should add info here about how to edit the default PSP to easily allow for `emptyDirs` everywhere. Doing this one-by-one for every deployment is cumbersome.

Author: added :heavy_check_mark:

- `cert-manager` must already be deployed in the cluster.
- it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
  - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.

> - with linkerd CNI enabled all pods linked by linkerd require a PodSecurityPolicy, which allows use of `EmptyDir`
>   - the injected linkerd-proxy container requires an `EmptyDir`

Author: @ced0ps 

```suggestion
- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
  - the injected linkerd-proxy container requires an `EmptyDir` volumes
```

>   - the injected linkerd-proxy container requires an `EmptyDir`

Reviewer 2: ```suggestion
- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
  - the injected linkerd-proxy container requires an `EmptyDir` volume
```
